internal/runtime/hcsv2/workload_container.go

```diff
@@ -163,8 +167,67 @@ func setupWorkloadContainerSpec(ctx context.Context, sbid, id string, spec *oci.
 	// Force the parent cgroup into our /containers root
 	spec.Linux.CgroupsPath = "/containers/" + id
 
+	if spec.Windows != nil && len(spec.Windows.Devices) > 0 {
+		if err := addNvidiaDevicePreHook(ctx, spec); err != nil {
+			return err
+		}
+	}
+
 	// Clear the windows section as we dont want to forward to runc
 	spec.Windows = nil
 
 	return nil
 }
+
+func addNvidiaDevicePreHook(ctx context.Context, spec *oci.Spec) error {
+	nvidiaToolBinary := "nvidiaPrestartHook"
+	nvidiaToolPath, err := exec.LookPath(nvidiaToolBinary)
+	if err != nil {
+		return errors.Wrapf(err, "failed to find %s for container GPU support", nvidiaToolBinary)
+	}
+
+	var args []string
+	args = append(args, nvidiaToolPath, "--debug=/nvidia-container.log", "--load-kmods", "--no-pivot", "--insecure-mode", "configure", "--ldconfig=/sbin/ldconfig")
+	if capabilities, ok := spec.Annotations["io.microsoft.container.gpu.capabilities"]; ok {
+		caps := strings.Split(capabilities, ",")
+		for _, c := range caps {
+			args = append(args, fmt.Sprintf("--%s", c))
+		}
+	}
+
+	for _, d := range spec.Windows.Devices {
+		switch d.IDType {
+		case "gpu":
+			args = append(args, fmt.Sprintf("--device=%s", d.ID))
```

Comment:
Isn't this the host vpci path?

Reply:
The reasoning behind this is [here](https://github.com/microsoft/hcsshim/pull/765#discussion_r373601465) I believe